tutorial/lesson_19_wrapper_funcs.cpp

```diff
@@ -0,0 +1,405 @@
+        //     h(x, y) = 3 + g(x, y) - f_in_h(x, y)
+
+        h.realize(5, 5);
+        // See figures/lesson_19_wrapper_unique.mp4 for a visualization.
+    }
+
+    {
+        // So far this may seem like a lot of pointless copying of
+        // memory. Func::in can be combined with other scheduling
+        // directives for a variety of purposes. The first we will
+        // examine is creating distinct realizations of a Func for
+        // several consumers and scheduling each differently.
+
+        // We'll start with nearly the same pipeline.
+        Func f("f_sched"), g("g_sched"), h("h_sched");
+        f(x, y) = x + y;
+        g(x, y) = 2 * f(x, y);
+        // h will use a far-away region of f
+        h(x, y) = 3 + g(x, y) - f(x + 93, y - 87);
+
+        // This time we'll inline f.
+        // f.compute_root();
+        g.compute_root();
+        h.compute_root();
+
+        f.in(g).compute_at(g, y);
+        f.in(h).compute_at(h, y);
+
+        // g and h now call f via distinct wrappers. The wrappers are
+        // scheduled, but f is not, which means that f is inlined into
+        // its two wrappers. They will each independently compute the
+        // region of f required by their consumer. If we had scheduled
+        // f compute_root, we'd be computing the bounding box of the
+        // region required by g and the region required by h, which
+        // would mostly be unused data.
+
+        // We can also schedule each of these wrappers
+        // differently. For scheduling purposes, wrappers inherit the
+        // pure vars of the Func they wrap, so we use the same x and y
+        // that we used when defining f:
+        f.in(g).vectorize(x, 4);
+        f.in(h).split(x, xo, xi, 2).reorder(xo, xi);
+
+        // Note that calling f.in(g) a second time returns the wrapper
+        // already created by the first call, it doesn't make a new one.
+
+        h.realize(8, 8);
+        // See figures/lesson_19_wrapper_vary_schedule.mp4 for a
+        // visualization.
+
+        // Note that because f is inlined into its two wrappers, it is
+        // the wrappers that do the work of computing f, rather than
+        // just loading from an existing computed realization.
+    }
+
+    {
+        // Func::in is useful to stage loads from a Func via some
+        // smaller intermediate buffer, perhaps on the stack or in
+        // shared GPU memory.
+
+        // Consider a pipeline that transposes some compute_root'd Func:
+
+        Func f("f_transpose"), g("g_transpose");
+        f(x, y) = sin(((x + y) * sqrt(y)) / 10);
+        f.compute_root();
+
+        g(x, y) = f(y, x);
+
+        // The execution strategy we want is to load an 4x4 tile of f
+        // into registers, transpose it in-register, and then write it
+        // out as an 4x4 tile of g. We will use Func::in to express this:
+
+        Func f_tile = f.in(g);
+
+        // We now have a three stage pipeline:
+        // f -> f_transpose -> g
```

Comment:
The naming here is confusingly at odds with naming the base f/g as "_transpose" for this example. The intermediate should at least be `f_tile`.

In general, I found the different naming of the base f/g/h Funcs in each case a bit odd, since:

1. The Func names do not actually match the simple base names used in the comments and pseudocode.
2. The names have to do with the example, not the role or use of the specific Func. So here, for example, there *is* a transpose going on, but "f_transpose" is not transposed.

I realize this is an effort to avoid cruft with "$unique" names at the global level, but I'm not sure it's the clearest choice. It seems like the code and comments matter more than what is going to get used inside the compiler or potentially printed if the reader modifies this file to do `print_loop_nest`, etc.